### PR TITLE
Implements improved weighted SSDeep match scoring

### PR DESCRIPTION
**Pull request by drewfarris**
* implements improved weighted SSDeep match scoring
* implements optional min score threshold parameter
* removed MATCH_RANK because it does not reflect sorted scoring

### CODE REVIEW

**Comment by drewfarris** [comment]
While the improved weighted scoring works fine, integration testing of the minScore threshold was not successful downstream. Investigation required. This could be merged as-is with a follow-up later to address the minScore issue.

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java [hunk 1]
```diff
@@ -1,48 +1,62 @@
...
             return false;
+
         NGramScoreTuple that = (NGramScoreTuple) o;
-        return ssDeepHash == that.ssDeepHash && Float.compare(that.score, score) == 0;
-    }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(ssDeepHash, score);
+        if (Float.compare(baseScore, that.baseScore) != 0)
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java]: Will epsilon be large enough here?

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java]: Actually, the only place this class is created is in the transformer and the values passed in are integers.  Why is baseScore and weightedScore stored as floats in the first place?

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/SSDeepHashScorer.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        }
+        
+        return (int) score;
+    }
+    
+    /**
+     * This is the low level chunk scoring algorithm. It takes two chunks and scores them on a scale of 0-100 where 0 is a terrible match and 100 is a great
+     * match. The chunkSize is used to cope with very small messages.
+     */
+    private static long scoreChunks(final String s1, final String s2, final long chunkSize) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/ssdeep/SSDeepHashScorer.java]: Should this be returning an int?

**Comment by drewfarris** [comment]
> While the improved weighted scoring works fine, integration testing of the minScore threshold was not successful downstream. Investigation required. This could be merged as-is with a follow-up later to address the minScore issue.

Disregard this - minScore is working properly downstream. The failure was somewhere between the chair and keyboard.

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java (hunk 1 above)
Comment (drewfarris) [reply warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java]: Refactoring artifact, will fix.

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/SSDeepHashScorer.java (hunk 2 above)
Comment (drewfarris) [reply warehouse/query-core/src/main/java/datawave/query/util/ssdeep/SSDeepHashScorer.java]: Likely another refactoring artifact, will fix. Thanks for the catch.

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java [hunk 3]
```diff
@@ -1,48 +1,65 @@
...
         return ssDeepHash;
     }
 
-    public float getScore() {
-        return score;
+    public float getBaseScore() {
+        return baseScore;
+    }
+
+    public float getWeightedScore() {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java]: return int?

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java [hunk 4]
```diff
@@ -1,48 +1,65 @@
 package datawave.query.util.ssdeep;
 
 import java.io.Serializable;
-import java.util.Objects;
 
 public class NGramScoreTuple implements Serializable, Comparable<NGramScoreTuple> {
     final SSDeepHash ssDeepHash;
-    final float score;
+    final float baseScore;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java:7]: int ?

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java [hunk 5]
```diff
@@ -1,48 +1,65 @@
...
+        this.weightedScore = weightedScore;
     }
 
     public SSDeepHash getSsDeepHash() {
         return ssDeepHash;
     }
 
-    public float getScore() {
-        return score;
+    public float getBaseScore() {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java:20]: int ?

File: warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java (hunk 4 above)
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/util/ssdeep/NGramScoreTuple.java:7]: Apparently this need to stay a float for future updates.

**Review by ivakegg — DISMISSED** [review]

**Review by apmoriarty — DISMISSED** [review]

**Review by keith-ratcliffe — APPROVED** [review]
Looks like only a minor test change since the prior approvals. I assume that change doesn't impact the overall status here

**Review by ivakegg — APPROVED** [review]

**Pull request merged**